### PR TITLE
Add logic to support testing with partial graph from build

### PR DESCRIPTION
**Pull request by mthalman**
These changes update Image Builder to provide support for the scenario described by #218.  The goal is for the test stage to be able to run tests for an image that was built by the build but the test requires images that were not built during the build.  An example of this is provided in #218.

To solve this, the image info files produced by the build are being utilized to track image tag information that can then be read by the test project to determine which images were actually produced by the build.  For those images, the test code will pull them from the staging location, where they were pushed during the build stage.  For any other images that the test requires, it will pull them from the public MCR location.

To make things cleaner, a new Image Builder command is defined which handles the merging of image info files (from each of the build jobs) into a single file.  This file can then be used by both the tests and the publish stage to do their work.  This command is called in a newly defined Post_Build stage that gets run after the Build stage.

### CODE REVIEW

File: Microsoft.DotNet.ImageBuilder/src/Commands/BuildCommand.cs [hunk 1]
```diff
@@ -110,6 +110,10 @@ private void BuildImages()
                         {
                             imageData.BaseImages = baseImageDigests;
                         }
+
+                        List<string> tags = platform.Tags.Select(tag => tag.Name).ToList();
+                        tags.Sort();
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Commands/BuildCommand.cs]: Curioius why not use OrderBy as part of the Linq expression?

File: Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.ImageBuilder.Models.Image;
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.ImageBuilder.Commands
+{
+    public class MergeImageInfoFilesCommand : Command<MergeImageInfoFilesOptions>
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs]: Thoughts on removing `Files` from the name?  It doesn't feel necessary from a conceptual level about what the command does and keeps the name of the command shorter.

File: Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        public override Task ExecuteAsync()
+        {
+            List<RepoData[]> srcReposList = Directory.EnumerateFiles(Options.SourceImageInfoFolderPath, "*.json", SearchOption.AllDirectories)
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs]: Nit - line length

File: Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                    $"No JSON files found in source folder '{Options.SourceImageInfoFolderPath}'");
+            }
+
+            List<RepoData> combinedRepos = new List<RepoData>();
+            foreach (RepoData[] repos in srcReposList)
+            {
+                ImageInfoHelper.MergeRepos(repos, combinedRepos);
+            }
+
+            string destinationContents = JsonHelper.SerializeObject(combinedRepos.OrderBy(r => r.Repo).ToArray()) + Environment.NewLine;
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs]: Nit - line length

File: Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesOptions.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+        public string DestinationImageInfoPath { get; set; }
+
+        public override void ParseCommandLine(ArgumentSyntax syntax)
+        {
+            base.ParseCommandLine(syntax);
+
+            string sourceImageInfoFolderPath = null;
+            syntax.DefineParameter(
+                "source-image-info-folder-path",
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesOptions.cs]: I realize this is the name already being used in `PublishImageInfoOptions` but it feels like a shorted name would be more consumer friendly.  How around just `source` and `destination` or `source-path` and `destination-path` ?

File: Microsoft.DotNet.ImageBuilder/src/Models/Image/ImageData.cs [hunk 6]
```diff
@@ -11,5 +11,6 @@ public class ImageData
     {
         [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
         public SortedDictionary<string, string> BaseImages { get; set; }
+        public List<string> Tags { get; set; } = new List<string>();
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/src/Models/Image/ImageData.cs]: I am wondering if this should be named more precisely.  These aren't the full set of tags as they don't include the shared tags.  What do you think about naming this `SimpleTags`?

File: Microsoft.DotNet.ImageBuilder/tests/ImageInfoHelperTests.cs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            List<RepoData> targetRepos = new List<RepoData>();
+            ImageInfoHelper.MergeRepos(repoDataSet, targetRepos);
+
+            CompareRepos(repoDataSet, targetRepos);
+        }
+
+        [Fact]
+        public void ImageInfoHelper_MergeRepos_ExistingTarget()
+        {
+            RepoData[] repoDataSet = new RepoData[]
```
Comment (MichaelSimons) [inline Microsoft.DotNet.ImageBuilder/tests/ImageInfoHelperTests.cs]: Would it make sense to share the image instances between the source repos and expected repos?

**Comment by mthalman** [comment]
I just realized I forgot to include the pipeline changes.  Added them now.

File: eng/common/templates/variables/docker-images.yml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 variables:
-  imageNames.imageBuilder.linux: mcr.microsoft.com/dotnet-buildtools/image-builder:debian-20190620175920
-  imageNames.imageBuilder.windows: mcr.microsoft.com/dotnet-buildtools/image-builder:nanoserver-20190620175920
+  imageNames.imageBuilder.linux: mthalman/test:imagebuilder
+  imageNames.imageBuilder.windows: mcr.microsoft.com/dotnet-buildtools/image-builder:nanoserver-20190606195358
```
Comment (MichaelSimons) [inline eng/common/templates/variables/docker-images.yml]: Because of this, I wonder if the build pipeline changes shouldn't be a separate PR?

File: eng/common/templates/stages/build-test-publish-repo.yml [hunk 9]
```diff
@@ -162,12 +162,18 @@ stages:
       dockerClientOS: windows
       buildJobTimeout: ${{ parameters.windowsAmdBuildJobTimeout }}
 
+- stage: Post_Build
```
[the remark this answers is not in the document]
Comment (mthalman) [reply eng/common/templates/stages/build-test-publish-repo.yml]: To me, `Finalize`  would imply the very last stage that would be run (i.e. after publish) which is not the case here.

File: Microsoft.DotNet.ImageBuilder/src/Commands/BuildCommand.cs (hunk 1 above)
Comment (mthalman) [reply Microsoft.DotNet.ImageBuilder/src/Commands/BuildCommand.cs]: Just wasn't thinking.  Changed to OrderBy.

File: Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs (hunk 2 above)
Comment (mthalman) [reply Microsoft.DotNet.ImageBuilder/src/Commands/MergeImageInfoFilesCommand.cs]: Makes sense.

File: eng/common/templates/jobs/build-images.yml [hunk 10]
```diff
@@ -40,36 +50,19 @@ jobs:
...
-        $(imageBuilder.queueArgs)"
-      displayName: Define imageBuilderBuildArgs Variable
   - script: >
       $(runImageBuilderCmd) build
       --manifest $(manifest)
       $(imageBuilderPaths)
       --os-type $(osType)
       --os-version "$(osVersion)"
       --architecture $(architecture)
+      --image-info-output-path "$(artifactsPath)/$(legName)-image-info.json"
```
[the remark this answers is not in the document]
Comment (mthalman) [reply eng/common/templates/jobs/build-images.yml]: Updated so it only gets set for internal builds.

File: eng/common/templates/stages/build-test-publish-repo.yml [hunk 11]
```diff
@@ -162,12 +162,18 @@ stages:
...
+  dependsOn: Build
+  condition: and(succeeded(), or(eq(variables['singlePhase'], ''), eq(variables['singlePhase'], 'build')))
```
[the remark this answers is not in the document]
Comment (mthalman) [reply eng/common/templates/stages/build-test-publish-repo.yml]: Updated so it only gets included for internal builds.

File: eng/common/templates/variables/docker-images.yml (hunk 8 above)
Comment (mthalman) [reply eng/common/templates/variables/docker-images.yml]: Agreed.  Reverting changes to the pipeline files for this PR.
